core/lib/server/pub.ts

```diff
@@ -541,55 +508,26 @@ export const createPubRecursiveNew = async <Body extends CreatePubRequestBodyWit
 			} as PubWithChildren;
 		}
 
-		const relatedPubs = await Promise.all(
-			Object.entries(body.relatedPubs).flatMap(async ([fieldSlug, relatedPubBodies]) => {
-				const field = pubFieldsForCommunity.find(({ slug }) => slug === fieldSlug);
-
-				if (!field) {
-					throw new NotFoundError(`No pub field found for slug '${fieldSlug}'`);
-				}
-
-				const relatedPubs = await Promise.all(
-					relatedPubBodies.flatMap(async ({ pub, value }) => {
-						const createdRelatedPub = await createPubRecursiveNew({
-							communityId,
-							body: pub,
-							trx,
-						});
-
-						if (!createdRelatedPub) {
-							throw new Error("Failed to create related pub");
-						}
-
-						const pubValue = await autoRevalidate(
-							trx
-								.insertInto("pub_values")
-								.values({
-									fieldId: field.id,
-									pubId: newPub.id,
-									value: JSON.stringify(value),
-									relatedPubId: createdRelatedPub.id,
-								})
-								.returningAll()
-						).executeTakeFirstOrThrow();
-
-						return {
-							...pubValue,
-							relatedPub: createdRelatedPub,
-						};
-					})
-				);
-
-				return relatedPubs;
-			})
-		);
+		// this fn itself calls createPubRecursiveNew, be mindful of infinite loops
+		const relatedPubs = await upsertPubRelations({
+			pubId: newPub.id,
+			relations: Object.entries(body.relatedPubs).flatMap(([fieldSlug, relatedPubBodies]) =>
+				relatedPubBodies.map(({ pub, value }) => ({
+					slug: fieldSlug,
+					value,
+					relatedPub: pub,
+				}))
+			),
+			communityId,
+			trx,
+		});
```

Comment:
i made `createPubRecursiveNew` use the new `upsertPubRelation` function, slightly DRYer